### DISCUSSION TITLE
Extend PrismaClient create() - trouble passing args

### DISCUSSION THREAD

**Discussion by schutzie** [prisma/prisma#22521] [Q&A]
### Question

Using the code below I'm attempting to automatically set two attributes: createdById and modifiedById.  The documentation for extending queries is pretty solid around select, but I can't find any examples for create, so I'm struggling here.  Below is the code and the error that I get.  In the error there's something about Without that I don't quite understand.  I struggle with creating records using a partial instance, like when there are 5 fields in a model and I only specify the required ones (omit the nullable fields).

So is my syntax for this correct?  And what's the issue with types here?  Thank you!
```js
const prisma = new PrismaClient()
  .$extends(
    {
      query: {
        item: {
          async create({model, operation, args, query})
          {
            args.data = {...args.data, createdById: "1", modifiedById: "1"}; // ERROR HERE

            // ALSO TRIED THIS WITH SIMILAR RESULT
            // args = { ...args, data: { ...args.data, createdById: "1", modifiedById: "1" } }

            return query(args);
          }
        }
      }
    }
  );
```

```
Type '{ createdById: string; modifiedById: string; ItemType?: undefined; Organization?: undefined; CreatedBy?: undefined; ModifiedBy?: undefined; itemId?: string | undefined; organizationId: string; ... 39 more ...; Transactions?: TransactionUncheckedCreateNestedManyWithoutItemInput | undefined; } | { ...; }' is not assignable to type '(Without<ItemCreateInput, ItemUncheckedCreateInput> & ItemUncheckedCreateInput) | (Without<...> & ItemCreateInput)'.
  Type '{ createdById: string; modifiedById: string; itemTypeItemTypeId?: undefined; organizationOrganizationId?: undefined; itemId?: string | undefined; organizationId: string; itemTypeId: string; ... 40 more ...; ModifiedBy: Prisma.UserCreateNestedOneWithoutModifiedByItemsInput; }' is not assignable to type '(Without<ItemCreateInput, ItemUncheckedCreateInput> & ItemUncheckedCreateInput) | (Without<...> & ItemCreateInput)'.
    Type '{ createdById: string; modifiedById: string; itemTypeItemTypeId?: undefined; organizationOrganizationId?: undefined; itemId?: string | undefined; organizationId: string; itemTypeId: string; ... 40 more ...; ModifiedBy: Prisma.UserCreateNestedOneWithoutModifiedByItemsInput; }' is not assignable to type 'Without<ItemCreateInput, ItemUncheckedCreateInput> & ItemUncheckedCreateInput'.
      Type '{ createdById: string; modifiedById: string; itemTypeItemTypeId?: undefined; organizationOrganizationId?: undefined; itemId?: string | undefined; organizationId: string; itemTypeId: string; ... 40 more ...; ModifiedBy: Prisma.UserCreateNestedOneWithoutModifiedByItemsInput; }' is not assignable to type 'Without<ItemCreateInput, ItemUncheckedCreateInput>'.
        Types of property 'ItemType' are incompatible.
          Type 'ItemTypeCreateNestedOneWithoutItemsInput' is not assignable to type 'undefined'.ts(2322)
(parameter) args: Prisma.ItemCreateArgs<InternalArgs & DefaultArgs>
```
The model for item is this.
```prisma
model Item {
  itemId                  String        @id @default(cuid())
  organizationId          String
  itemTypeId              String
  title                   String        @db.VarChar(255)
  ...
  createdById                String @db.VarChar(128)
  modifiedById               String @db.VarChar(128)
  CreatedBy                  User   @relation("itemCreatedBy", fields: [createdById], references: [id])
  ModifiedBy                 User   @relation("itemModifiedBy", fields: [modifiedById], references: [id])
  ...
}
```
And for user is this.
```prisma
model User {
  id    String @id @default(cuid())
  email String @unique
  ...
  CreatedByItems                 Item[]                @relation("itemCreatedBy")
  ModifiedByItems                Item[]                @relation("itemModifiedBy")
  ...
}
```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @schutzie 👋 

Thank you for raising this question.

The arguments you tried modifying `createdById` and `modifiedbyId` are foreign key fields. Looking at the `ItemCreateInputType`, these are the allowable fields you can modify in `args` object
```ts
export type ItemCreateInput = {
    itemId?: string
    organizationId: string
    itemTypeId: string
    title: string
    CreatedBy: UserCreateNestedOneWithoutCreatedByItemsInput
    ModifiedBy: UserCreateNestedOneWithoutModifiedByItemsInput
  }
```

you can have something like:
```js
const prisma = new PrismaClient({
  log: ['query']
}).$extends(
  {
    query: {
      item: {
        async create({model, operation, args, query})
        {
          
          args.data = {...args.data,  itemTypeId: '1',organizationId: '1', title: 'Test Item'}; 

          return query(args);
        }
      }
    }
  }
);
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by schutzie** [reply]
This is so much further than I've been able to get on my own. Thank you! So my follow up question is two parts.

1. I do not have to create the UserCreateNestedOneWithoutCreatedByItemsInput and UserCreateNestedOneWithoutModifiedByItemsInput types/classes, correct? These are generated types?
2. This is great, except that the Item model contains far more fields than just the ones I shared. There are probably 3-4 dozen. There are probably another 2-3 dozen models where I want to use these two CreatedBy and ModifiedBy attributes. Is there a better (shorter) way around creating this for every one of my models that contain those two attributes? Maybe some kind of generic type based on the concept of a model, not for a specific model?

Thank you so much for your reply. This is definitely getting me on the right track. Just looking for something a little lower maintenance.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
